### PR TITLE
Population mode

### PR DESCRIPTION
**Pull request by tbekolay**
This is a migration of https://github.com/nengo/nengo/issues/67 with some additional context and prose. Like #2, it depends on #1.

### CODE REVIEW

File: 001-population-mode.rst [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+Context
+=======
+
+Nengo can use several neuron types,
+as well as a special type, ``nengo.Direct``.
+Direct mode, as it's commonly called,
+represents vectors perfectly
+and transforms them through normal Python functions
+rather than neural approximations.
+This is faster than using neurons,
```
Comment (jgosmann) [inline 001-population-mode.rst:23]: This is not necessarily true.

Comment (tbekolay) [reply 001-population-mode.rst:23]: Fair enough, I'll add "generally" or "usually".

**Comment by AllenHW** [comment]
Just wanted to give an update on this as the approach is becoming more clear. Below is one approach based on Bryan Tripp's paper.  

**Summary of the paper**
Each neural population (i.e. ``Ensemble``) computes a function *f(x)*. This neural function is different from the ideal function *F(x)*, and we approximate the the difference *f(x) - F(x) = e(x)* as *e(x) = b(x) + s(x)* where *b(x)* stands for *bias* and *s(x)* for *noise*. We use different interpolation methods to approximate *b(x)*, and a time series model called ARMA to approximate *s(x)*

**Implementation Approach**
 - We can obtain *b(x)* by taking the difference between the ideal function *F(x)* and decoded function without the noise. We can obtain *F(x)* either using Direct mode, or, given an ``Ensemble`` and a ``Connection``, we know what the ideal function is. We can obtain decoded function without noise by multiplying neural activities and the decoder (``tuning_curve()`` gets us neural activities). 
- We can obtain *s(x)* by taking the difference between the the decoded function without noise and the actual output of the network. 
- After we obtain models for *b(x)* and  *s(x)*, we can approximate the neural output as *F(x) + b(x) + s(x)*

Note that the current approach requires building the network as well as simulating it. However, we only need to simulate the network for a short period of time

**Comment by tbekolay** [comment]
One question I have about that description is: does the function `f(x)` belong to the ensemble or to the connection? I ask because implementation wise, I would think of the two very differently.

If f(x) belongs to the ensemble, then `f(x)` should be the identity function, meaning that you're determining, for each ensemble, how good that ensemble is at representing `x`. I.e., for all parts of the state space of `x`, we're finding `b(x)` and `s(x)` such that we can tell what the encoded `x` would be without having to simulate all the neurons again (because we assume if we did that, it would be the same as when we did it the last time). This would mean that connections from that ensemble would essentially compute `f(x + b(x) + s(x))`.

If f(x) belongs to the connection, then `f(x)` would be the actual function that the user asked for, meaning that you're determining, for each connection, how good that ensemble is at computing `f(x)`. I.e., for all possible input values `x`, we're finding `b(x)` and `s(x)` such that we can tell what the computed `f(x)` would be without having to simulate all the neurons again. This would mean that connections from the ensemble would essentially compute `f(x) + b(x) + s(x)`.

I don't have access to the original paper to confirm myself, plus I think the discussion is helpful to have :)

**Comment by AllenHW** [comment]
Sorry for the super late reply... I missed the comment.

f(x) belongs to the connection -  and f(x) is what the user has asked. And you're right that we're basically finding b(x) for each x by selecting some evaluation points. For the noise term s(x), we would characterize the noise by its amplitude, frequency, etc, since dependence on x is random.

**Comment by tbekolay** [comment]
Migrated to a Google Doc.

**Pull request closed without merging**